ethcore/src/engines/mod.rs

```diff
@@ -338,6 +340,16 @@ pub trait Engine<M: Machine>: Sync + Send {
 	fn is_timestamp_valid(&self, header_timestamp: u64, parent_timestamp: u64) -> bool {
 		header_timestamp > parent_timestamp
 	}
+
+	/// Check whether a given block is the best block.
+	fn is_new_best(&self, bytes: &[u8], best_block_total_difficulty: U256, provider: &BlockProvider) -> bool {
```

Comment:
can we abstract around `total_difficulty` and `provider` at the `Machine` level?

Reply:
Will give it a try.

Reply:
Here's an attempt: I added assoc type `Machine::BlockMetadata` and `Machine::BlockProvider`, and make `Engine::is_new_best` depend on them only. For `EthEngine`, `BlockMetadata` is currently just `U256` which is the total difficulty (and we can extend it to contain `Vec<(Bytes, Bytes)>` in the future).

Client would first call `Engine::generate_metadata` to get the current set of metadata, and then call `Engine::is_new_best` before passing it to blockchain storage.